fdbclient/include/fdbclient/TenantManagement.actor.h

```diff
@@ -174,48 +193,50 @@ Future<TenantMapEntry> createTenant(Reference<DB> db, TenantName name) {
 }
 
 ACTOR template <class Transaction>
-Future<Void> deleteTenantTransaction(Transaction tr, TenantNameRef name) {
-	state Key tenantMapKey = name.withPrefix(tenantMapPrefix);
-
+Future<Void> deleteTenantTransaction(Transaction tr,
+                                     TenantNameRef name,
+                                     Optional<int64_t> tenantId = Optional<int64_t>()) {
 	tr->setOption(FDBTransactionOptions::RAW_ACCESS);
 
-	state Optional<TenantMapEntry> tenantEntry = wait(tryGetTenantTransaction(tr, name));
-	if (!tenantEntry.present()) {
-		return Void();
-	}
+	state Future<Optional<TenantMapEntry>> tenantEntryFuture = tryGetTenantTransaction(tr, name);
+	wait(checkTenantMode(tr));
 
-	state typename transaction_future_type<Transaction, RangeResult>::type prefixRangeFuture =
-	    tr->getRange(prefixRange(tenantEntry.get().prefix), 1);
-	RangeResult contents = wait(safeThreadFutureToFuture(prefixRangeFuture));
-	if (!contents.empty()) {
-		throw tenant_not_empty();
-	}
+	state Optional<TenantMapEntry> tenantEntry = wait(tenantEntryFuture);
+	if (tenantEntry.present() && (!tenantId.present() || tenantEntry.get().id == tenantId.get())) {
```

Comment:
This is to maintain the idempotency property. The contract of `deleteTenantTransaction` (which I could document) is that it will delete the tenant if it exists, and if it doesn't it will assume that it already succeeded in deleting it. If you want to ensure that the tenant is there, you would need to do so externally as part of the same transaction.

If the ID has changed, this is a special case of the above. It could be that you successfully deleted the tenant, got an unknown result, and then someone else recreated it before you retried. In that case you proceed in the same way.

We could optionally return a `bool`, but no callers are using it. Because this is an ACTOR, it's a little bit annoying to ignore a return value.

Reply:
I should also add that this function is used internally and is not exposed as such to users. The main reason for adding the argument is in the metacluster case, where we cannot delete the tenant transactionally. In that case, we don't want to have a weird interleaving sequence something like:

1. Mark tenant X in removing state with ID 1 from management cluster (attempt A)
2. Mark tenant X in removing state with ID 1 from management cluster (attempt B)
3. Delete tenant X from data cluster with ID 1 (attempt B)
4. Delete tenant X with ID 1 from management cluster (attempt B)
5. Create tenant X with ID 2 on management cluster
6. Create tenant X with ID 2 on data cluster
7. Delete tenant X from data cluster with ID 1 (attempt A) (i.e. the second phase of attempt A got delayed)

In this case, we want step 7 to be a no-op, so it sends the ID of the tenant it expects to be operating on and does nothing if the ID doesn't match. This logic is all internal, though.